File: 121BestTimetoBuyandSellStock.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

class Solution {
    public:
        int maxProfit(vector<int>& prices) {
            int min_price = INT_MAX;
            int max_profit = 0;
            for (int price: prices){
                if (price<min_price){
                    min_price = price;
                }
                int profit = price - min_price;
                if (profit>max_profit){
                    max_profit = profit;
                }
            }
            return max_profit;
        }
    };

int main(){
    Solution sol;
    vector<int> prices = {7,1,5,3,6,4};
    int ans = sol.maxProfit(prices);
    cout << ans << endl;
}

File: 121_best_time_to_buy_and_sell_stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        min_prices = prices[0] # 紀錄最低進場價
        profit = 0
        max_profit = 0
        for price in prices:
            if price<min_prices:
                min_prices = price # 更新最低價
            else:
                profit = price - min_prices # 非最低價，賣出記錄獲利
            max_profit = max(max_profit, profit) # 更新最高獲利
        return max_profit
        

        
if __name__ == "__main__":
    sol = Solution()
    prices = [7,1,5,3,6,4]
    print(sol.maxProfit(prices))